sklearn/linear_model/sag_fast.pyx

```diff
@@ -32,49 +27,240 @@ cdef inline double fmax(double x, double y) nogil:
+        np.zeros(n_classes, dtype=np.double, order="c")
+    cdef double* gradient = <double*> gradient_array.data
+
+    with nogil:
+        for i in range(n_samples):
+            # get next sample on the dataset
+            dataset.next(&x_data_ptr, &x_ind_ptr, &xnnz,
+                         &y, &sample_weight)
+
+            # prediction of the multinomial classifier for the sample
+            predict_sample(x_data_ptr, x_ind_ptr, xnnz, weights, wscale,
+                           intercept, prediction, n_classes)
+
+            # compute the gradient for this sample, given the prediction
+            multiloss._dloss(prediction, y, n_classes, sample_weight, gradient)
+
+            # compute the loss for this sample, given the prediction
+            sum_loss += multiloss._loss(prediction, y, n_classes, sample_weight)
+
+            # update the sum of the gradient
+            for j in range(xnnz):
+                f = x_ind_ptr[j]
+                val = x_data_ptr[j]
+                for c in range(n_classes):
+                    sum_gradient[f * n_classes + c] += gradient[c] * val
+
+    return sum_loss, sum_gradient_array
+
+
 def sag(SequentialDataset dataset,
-        np.ndarray[double, ndim=1, mode='c'] weights_array,
-        double intercept,
+        np.ndarray[double, ndim=2, mode='c'] weights_array,
+        np.ndarray[double, ndim=1, mode='c'] intercept_array,
         int n_samples,
         int n_features,
+        int n_classes,
         double tol,
         int max_iter,
-        LossFunction loss,
+        str loss_function,
         double step_size,
         double alpha,
-        np.ndarray[double, ndim=1, mode='c'] sum_gradient_init,
-        np.ndarray[double, ndim=1, mode='c'] gradient_memory_init,
+        np.ndarray[double, ndim=2, mode='c'] sum_gradient_init,
+        np.ndarray[double, ndim=2, mode='c'] gradient_memory_init,
         np.ndarray[bint, ndim=1, mode='c'] seen_init,
         int num_seen,
         bint fit_intercept,
-        double intercept_sum_gradient,
+        np.ndarray[double, ndim=1, mode='c'] intercept_sum_gradient_init,
         double intercept_decay,
         bint verbose):
+    """Stochastic Average Gradient (SAG) solver.
 
-    # true if the weights or intercept are NaN or infinity
-    cdef bint infinity = False
-    # the pointer to the coef_ or weights
-    cdef double* weights = <double * >weights_array.data
+    Used in Ridge and LogisticRegression.
+
+    Reference
+    ---------
+    Schmidt, M., Roux, N. L., & Bach, F. (2013).
+    Minimizing finite sums with the stochastic average gradient
+    https://hal.inria.fr/hal-00860051/PDF/sag_journal.pdf
+    (section 4.3)
+    """
     # the data pointer for X, the training set
-    cdef double *x_data_ptr
+    cdef double *x_data_ptr = NULL
     # the index pointer for the column of the data
-    cdef int *x_ind_ptr
-    # the label for the sample
+    cdef int *x_ind_ptr = NULL
```

Comment:
I used the same name as in `seq_dataset`, which is related to classical names for scipy sparse CSR structure.